File: homework1/src/nearest.py
#!/usr/bin/env python

import rospy
from sensor_msgs.msg import LaserScan
from std_msgs.msg import String

pub = rospy.Publisher('value_node', String, queue_size=10)


def sub_callback(data):
	min_value = 'Min value = ' + str(min(data.ranges))
	rospy.loginfo('%s', min_value)
	pub.publish(min_value)

def nearest():
	rospy.init_node('nearest', anonymous=True)
	rospy.Subscriber("/base_scan", LaserScan, sub_callback)
	rospy.spin()
	
if __name__ == '__main__':
    try:
        nearest()
    except rospy.ROSInterruptException:
        pass

File: src/hw1/src/nearest.cpp
#include "ros/ros.h"
#include "std_msgs/String.h"
#include "sensor_msgs/LaserScan.h"

ros::Publisher pub;

void callback(const sensor_msgs::LaserScan::ConstPtr& data){
	double min_value = DBL_MAX;
    int length = (data->ranges).size();
    int i = 0;
    
    while(i < length) {	
		double curr_value = (data->ranges)[i];
		if(curr_value < min_value) min_value = curr_value;
		i++;
    }
	
	if(ros::ok()) {
		std::stringstream ss;
		ss << "Nearest value is: " << min_value;
		
		std_msgs::String msg;
		msg.data = ss.str();
		
		ROS_INFO("%s", msg.data.c_str());
		pub.publish(msg);
	}
}

int main(int argc, char **argv) {
	ros::init(argc, argv, "nearest");

	ros::NodeHandle nodeH;
		
	pub = nodeH.advertise<std_msgs::String>("nearest_value", 1000);
	ros::Subscriber sub = nodeH.subscribe("/base_scan", 1000,  callback);

	ros::spin();
	return 0;
}
